### ISSUE TITLE
AiModelResult.Predict returns uniform-zero output even when underlying Model.Predict returns trained predictions

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1267]
## Severity

**Critical** — every consumer using the AiModelBuilder facade for inference receives non-functional predictions despite training succeeding.

## Symptom (concrete numbers)

Same trained Transformer, two prediction paths, two completely different results:

- **Via facade**: \`_trained.Predict(input)\` returns logits with argmax always at 0 (null byte) → top-1 = **0.00%**, PPL = **256.00** (= V exactly = uniform random)
- **Via direct model**: \`_trainedModel.Predict(input)\` returns logits with argmax tracking unigram prior (most common byte = space) → top-1 = **13.40%**, PPL = **242.42**

Both calls operate on the SAME trained Transformer instance after the same training run. The facade path actively corrupts inference.

## Reproducer setup

\`\`\`csharp
var model = new Transformer<float>(arch, lossFunction: new CategoricalCrossEntropyLoss<float>());
// ... train via builder.BuildAsync() OR via model.TrainBatched(...) ...

// Facade path — returns uniform-zero
var trainedResult = builder.BuildAsync().GetAwaiter().GetResult();
var p1 = trainedResult.Predict(input);  // logits all ≈ 0, argmax = 0 forever

// Direct path — returns trained predictions
model.SetTrainingMode(false);
var p2 = model.Predict(input);  // logits track unigram, argmax = most-common byte
\`\`\`

The direct path goes through \`Transformer.Predict\` → \`ForwardForTraining(false)\` and returns sensible output. The facade path goes \`AiModelResult.Predict → Model.Predict(normalizedNewData)\` (per src/Models/Results/AiModelResult.cs:1919) and somehow loses the trained state.

## Suspected root cause locations (in priority order)

1. **JitCompiledFunction capture**. \`AiModelBuilder.BuildAsync\` calls \`BuildCompiledPredictFunction(optimizationResult.BestSolution)\` at the END of the build (lines 1574/1987/3213/3706). If the JIT trace captures parameters BY VALUE (snapshot) rather than BY REFERENCE, post-trace param updates won't be reflected — but that wouldn't explain it because we set \`ConfigureJitCompilation(JitCompilationConfig.Disabled)\` and the bug still reproduces.

2. **AiModelResult.Predict's preprocessing branch** at \`AiModelResult.cs:1864\`:
\`\`\`csharp
var normalizedNewData = PreprocessingInfo?.IsFitted == true
    ? PreprocessingInfo.TransformFeatures(dataForPrediction)
    : dataForPrediction;
\`\`\`
   If \`PreprocessingInfo\` gets a default-fitted state somewhere even when no pipeline was configured, \`TransformFeatures\` could mangle the input.

3. **\`ApplySelectedFeaturesForPrediction\`** at line 1872. Feature-selection metadata could be applied even when training didn't use feature selection, slicing the input incorrectly.

4. **\`EnsureStatelessInferenceOptimizationsInitialized\`** at line 1882. Even though \`InferenceOptimizationConfig == null\` for our case, this path may be entering inadvertently and creating a fresh untrained copy.

5. **Param sharing via ParameterBuffer**. Training writes through buffer-aliased views; inference may read from un-aliased original tensors that were never updated. Possible if \`RestoreOriginalParameters\` (line 3724) is missing a copy direction.

## Severity / blast radius

Every example in the README that uses the \`AiModelBuilder.BuildAsync\` → \`result.Predict\` flow is broken on neural-network models. This is the documented happy path:

\`\`\`csharp
// docs/api-keys.md, README.md, examples/Transformer.cs all show:
var result = await new AiModelBuilder<float, Tensor<float>, Tensor<float>>()
    .ConfigureModel(model)
    .ConfigureOptimizer(optimizer)
    .ConfigureDataLoader(loader)
    .BuildAsync();
var prediction = result.Predict(input);   // ← BROKEN: returns uniform output
\`\`\`

## Workaround until fixed

Bypass the AiModelResult wrapper entirely:

\`\`\`csharp
// Train via direct model API (TrainBatched added in PR #1265)
model.TrainBatched(inputs, targets);
model.SetTrainingMode(false);
var prediction = model.Predict(input);  // works — returns trained output
\`\`\`

This is what HarmonicEngine's predictor at \`tests/Learning/Support/AiDotNetFacadeTransformerLMPredictor.cs\` now does as the workaround (commit landing this hour on PR #1265 sister branch).

## Related

- ooples/AiDotNet#1264 (Transformer convergence) — fixed by PR #1265's Adam-default + TrainBatched + facade-step rewire. This issue is a SEPARATE bug exposed by those fixes; the model now trains correctly but inference through the facade is broken.

## Bar for fix

Add to \`NeuralNetworkModelTestBase\` (PR #1265 already strengthened it with two new invariants): \`AiModelBuilderFacade_Predict_MatchesDirectModelPredict\`. After a build + training run, assert \`result.Predict(input).MaxAbsDiff(model.Predict(input)) < 1e-5\`. That's the single test that would have caught this.

**Comment by ooples** [comment]
## Root cause found and fixed (PR #1265, commit e81b7ef52)

**The bug was NOT in `AiModelResult.Predict` — it was in `AiModelBuilder.BuildSupervisedInternalAsync` at the regular-training path** (`src/AiModelBuilder.cs` ~line 3022):

```csharp
optimizationResult = finalOptimizer.Optimize(optimizationInputData);
```

For neural-network models, `finalOptimizer.Optimize` runs the OUTER clone-evaluate-select hyperparameter search loop — it **never** invokes the tape-based per-iteration training path that actually updates attention / FFN / embedding weights. The `OptimizationResult.BestSolution` it returns is a fresh CLONE whose weights are at their initial Xavier state — NOT the user's trained model.

This produced the symptom in #1267: the model the user passed to `ConfigureModel(model)` and the `result.Model` field returned by `BuildAsync()` were **two different `Transformer<T>` instances**:

- The user's reference: gets trained if the user invokes `model.Train` directly (works fine).
- The result wrapper's `Model`: never trained, all-zero `Layers[0]` params → `Predict` outputs the uniform softmax.

Diagnostic trace from a failing run (added then removed in this commit):
```
[CTOR] AFTER assign Model = BestSolution: Layers[0].Params=0,0,0,0  (instance hash 45788687)
[TEST-PRE] direct-call model.Layers[0].Params=0.0099,-0.0648,...    (instance hash 28068188 — user's ref, has Xavier init)
[TEST-PRE] result.Model is same? False
```

After fix:
```
[CTOR] AFTER assign Model = BestSolution: Layers[0].Params=0.0099,-0.0648,...  (instance hash 28068188)
[TEST-PRE] result.Model is same? True
```

### Fix

In `BuildSupervisedInternalAsync`'s regular-training branch, detect `INeuralNetwork<T>` + `Tensor<T>` inputs and route to `nn.Train(xTrainTensor, yTrainTensor)` directly. `nn.Train` dispatches through `TrainWithTape → Optimizer.Step(TapeStepContext)` which IS the supported path for tape-based training (uses configured Adam / AdamW / SGD-with-momentum moments). The result wrapper's `BestSolution` is then set to the trained model itself (same reference the user passed in), with identity `SelectedFeatureIndices` so `ApplySelectedFeaturesForPrediction` no-ops.

The non-NN path (linear regression, decision trees, SVM, ensembles, …) still calls `finalOptimizer.Optimize` unchanged — that's the correct optimizer-driven path for those families.

### Verification

New invariant test (`AiModelBuilderFacadePredictParityTests.Facade_Predict_MatchesDirectModelPredict_AfterBuildAsync`) verifies `maxAbsDiff(facade_predict, direct_predict) < 1e-3` after `BuildAsync`. **Passes after fix, fails before.** This is one of 6 integration tests added to PR #1265 to prevent regression.

### What was also fixed earlier in this PR

While digging for this root cause we found and fixed several adjacent bugs in the same training path; full list in PR #1265 description:

1. **Default optimizer was vanilla SGD, not Adam** — fixed.
2. **OptimizerBase counter inversion** (improving fitness wrongly incremented `IterationsWithImprovement`'s opposite branch) — fixed.
3. **`AiModelResult.Predict` didn't call `SetTrainingMode(false)`** before `Model.Predict` — fixed.
4. **`AiModelBuilder` data-loader path bypassed the optimizer** (used legacy `ComputeGradients/ApplyGradients` per-sample instead of `nn.Train(batched)`) — fixed.

Closing this issue. Will close once PR #1265 merges.
